### PR TITLE
Feature lwdev 8682 avoid candle copies in cache

### PR DESCRIPTION
**Pull request by KonstantinRyazantsev**

### CODE REVIEW

File: src/Lykke.Job.CandlesHistoryWriter.Services/Candles/CandlesCacheInitalizationService.cs [hunk 1]
```diff
@@ -47,6 +50,13 @@ public async Task InitializeCacheAsync()
...
             // the invalidation timer fires before the cache loading has stopped. This will be a signal 
             // to skip timer-based invalidation.
+            // Below we combine two approaches:
+            // - we're exporting a fast signal for any timer-based routine that we have already been initializing the cache;
+            // - and we additionally block all the cache-related operations for other parts of code.
+            // The first is needed to avoid queueing of timer events. If we simply use blocks, we will finally face a problem
+            // when cache initialization may become infinitly repeated. The second is important for avoidining a multi-threaded
+            // write operations to the cache: if we've got a candle update set, we need to await for cache fill completion and
+            // then proceed.
             lock (_initializationStateLocker)
```
Comment (MikhailGerasimov) [inline src/Lykke.Job.CandlesHistoryWriter.Services/Candles/CandlesCacheInitalizationService.cs]: Do not use lock inside async. It blocks threads from the thread pool and can lead to thread pool thread starvation

File: src/Lykke.Job.CandlesHistoryWriter.Services/Candles/CandlesCacheSemaphore.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        public void Release()
+        {
+            _sem.Release();
+        }
+
+        public async Task WaitAsync()
+        {
+            await _sem.WaitAsync();
```
Comment (MikhailGerasimov) [inline src/Lykke.Job.CandlesHistoryWriter.Services/Candles/CandlesCacheSemaphore.cs]: async await redundant

File: src/Lykke.Job.CandlesHistoryWriter.Services/Candles/CandlesCacheInitalizationService.cs (hunk 1 above)
Comment (ozhelenkov) [reply src/Lykke.Job.CandlesHistoryWriter.Services/Candles/CandlesCacheInitalizationService.cs]: missed that as was not introduced in that PR. Good catch

Comment (KonstantinRyazantsev) [reply src/Lykke.Job.CandlesHistoryWriter.Services/Candles/CandlesCacheInitalizationService.cs]: Fair remark. Will add the task to fix it

**Pull request merged**